File: 2022/Day7.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def gettree(tree, path):
    for p in path:
        tree = tree[p]
    return tree

def parse(inp):
    path = ['/']
    tree = {'/': {}}
    for l in lines(inp)[1:]:
        s = l.split()
        if l.startswith("$"):
            if l.startswith("$ cd"):
                cd = s[2]
                if cd == "..":
                    path.pop()
                else:
                    path.append(cd)
        else:
            a, b = s
            if a == 'dir':
                gettree(tree, path)[b] = {}
            else:
                gettree(tree, path)[b] = int(a)

    return tree

def solve(tree):
    p1 = 0
    sizes = []
    def size(tree):
        nonlocal p1, sizes
        s = 0
        for v in tree.values():
            if isinstance(v, int):
                s += v
            else:
                s_v = size(v)
                s += s_v
                sizes.append(s_v)

        if s <= 100000:
            p1 += s

        return s

    total = size(tree)
    needed = 30000000 + total - 70000000
    p2 = min(x for x in sizes if x >= needed)
    return (p1, p2)

def main():
    inp = parse(data())
    p1, p2 = solve(inp)
    print(f"Solution for part 1:\n{p1}")
    print(f"Solution for part 2:\n{p2}")

if __name__ == "__main__":
    main()

# Solution part 1: 1453349
# Solution part 2: 2948823
# Leaderboard: 954 / 1525

File: 2022/Day7.hs
#! /usr/bin/env -S runhaskell
import Text.Read (readMaybe)

data File = File Int | Dir [File] deriving (Show, Read)

type Input = [Int]
type Output = Int

parse :: String -> Maybe Input
parse = (snd . getSizes <$>) . readMaybe . removeTrailing . balance 0 . transform . lines
  where
    transform [] = ""
    transform ("$ ls":ls)                  = transform ls
    transform (('d':'i':'r':' ':_):ls)     = transform ls
    transform ("$ cd ..":ls)               = "]," ++ transform ls
    transform (('$':' ':'c':'d':' ':_):ls) = "Dir [" ++ transform ls
    transform (file:ls)                    = let [x, _] = words file; in "File " ++ x ++ ',' : transform ls
    balance l [] = concat $ replicate l "],"
    balance l (x:xs) = x : balance (case x of '[' -> l+1; ']' -> l-1; _ -> l) xs
    removeTrailing (x:y:xs) = if [x,y] == ",]" then y : removeTrailing xs else x : removeTrailing (y:xs)
    removeTrailing (",") = []
    getSizes (File n) = (n, [])
    getSizes (Dir xs) = let (sums, sizes) = unzip $ map getSizes xs in ((sum sums), (sum sums) : (concat sizes))

part1 :: Input -> Output
part1 = sum . filter (<=100000)

part2 :: Input -> Output
part2 sizes = let thres = (maximum sizes) - 40000000 in minimum $ filter (>=thres) sizes

main :: IO()
main = do
    f <- readFile("7.in")
    let input = parse f
    case input of
      Nothing -> putStrLn "Unable to read input."
      Just i -> do
        putStr "Solution for part 1: "
        putStrLn . show $ part1 i
        putStr "Solution for part 2: "
        putStrLn . show $ part2 i

-- Solution for part 1: 1453349
-- Solution for part 2: 2948823
